### ISSUE TITLE
[11.0] ERROR: could not serialize access due to concurrent update

### ISSUE THREAD

**Issue by mamiu** [odoo/odoo#23086]
**Impacted versions:**
11 (and probably below)

**Steps to reproduce:**
Lets assume you have a module called `my_module` which contains a model `sample`.
Now create a controller with the following code:

```python
import base64
from odoo import http

@http.route(['/upload-file'], type='http', auth='none', methods=['POST'], csrf=False)
def upload_file(self, **kwargs):

    sample_id = http.request.env['my_module.sample'].sudo().search([], limit=1)

    if len(sample_id > 0):
        sample_id.filename = kwargs['fileName']
        sample_id.file = base64.encodebytes(kwargs['fileData'].read())
```
(I know this code is not clean code, it's just to reproduce the error)

**Current behavior:**
If this `/upload-file` API is getting called and there are no other calls to this API at the same time, everything works like expected.
Here's an example how this API could be called from javascript:
```html
<input type="file" name="file-field" onchange="handleFileChange" />
```
```js
function handleFileChange(event) {
  const file = event.target.files[0];
  let formdata = new FormData();

  formdata.append('fileName', file.name);
  formdata.append('fileData', file);

  let xhr = new XMLHttpRequest();
  let url = encodeURI('/upload-file');
  xhr.open('POST', url);
  xhr.send(formdata);
}
```

But if this API is getting called in parallel, on the server following error occurs:
```
2018-02-15 07:27:33,784 1 INFO test2 odoo.service.model: SERIALIZATION_FAILURE, retry 1/5 in 0.0883 sec...
ERROR: could not serialize access due to concurrent update
```

This error is also thrown, if the simultaneously uploaded files are saved to different fields on the same record.

**Expected behavior:**

As long as the API method saves the uploaded files to different fields, there shouldn't be any database update conflicts.

@Yenthe666 thanks in advance for your help! 👍

**Comment by Yenthe666** [comment]
@JKE-be this example is to reproduce it from controllers/website function so you're probably the best one to have a look at this.
@xmo-odoo and @odony you've both been looking at and handled bug reports about serialize access so I guess you two can give good feedback here?

**Comment by willmakenoise** [comment]
I'm seeing the same issue on Odoo 10

**Comment by sacostaylara** [comment]
Any solution for this issue?

**Comment by phoenixsampras** [comment]
The same happens with POS 
ERROR:  could not serialize access due to concurrent update

This in an old issue, please help!!

**Comment by dougngo** [comment]
Anybody? Still waiting on a solution here as well

**Comment by bnisevic** [comment]
I'm getting this too on Odoo v12 EE.

**Comment by legalsylvain** [comment]
Same problem here. Point of sale / V8.0. (but it seems unrelated to the version).
Original issue : https://github.com/odoo/odoo/issues/18546

**Comment by bnisevic** [comment]
Maybe it's an Odoo feature and we are not aware of it.

**Comment by legalsylvain** [comment]
Hi

> provide detailed steps to reproduce the problem.

I'd like to, but I can not reproduce the error with demo data, because the problem depends on the activity I guess. For my side, it occures when the create_from_ui (point_of_sale) function is called many times.

>  in a new issue if it happens on another area.

Well, the issue is generic, as it describe an error that occures when calling simultaneously the same function via API.

context : 

- Odoo 8.0 (OCB)
- 47 companies
- 45 Point Of Sale (Pos Config)
- ~3000 pos order per day. maximum of ~300 per hour.

PostgreSQL 9.5.12

Error : 

```
TransactionRollbackError could not serialize access due to concurrent update
bad query: UPDATE "account_bank_statement" SET "total_entry_encoding"='198.00' WHERE id = 54913
```

My comment is not precise, and I'm sorry for that, but I don't think it's off-topic.

kind regards.

**Comment by pedrobaeza** [comment]
@legalsylvain my comment was not directly pointed to you, but it's also related. As said, a concurrent update is just that: 2 things (workers or threads) updating the same record at the same time. And thus, this applies to any part of the system.

If this can be reproducible with just standard core, then can be considered an Odoo bug, and should be put in a different issue with steps to reproduce it. Take into account also that 8.0 is not a supported version anymore.

If not, it's just simply luck that 2 users/bots matches in the same moment for the same record. Odoo has already done some actions for avoiding this on normal operation (disable cron execution while updating/installing modules), but not too much can be done in the rest, except retrying the operation when happening.

**Comment by odony** [comment]
For the record, the issue described by the OP looks invalid, or rather it's the expected behavior. AFAICS that sample controller searches for the first `my_module.sample` record and updates it. If this is done in multiple threads in parallel it *must* create a concurrent update transaction conflict : the database cannot guarantee what should be the final state of the record.
Obviously such a controller would make no sense, and should be writing on different records. I'd rather close this issue because it's misleading to view this expected behavior as a bug.

**Comment by bnisevic** [comment]
You can offtopicing me as much as you want but this is Odoo issue since default isolation level change and it is present even now v12.

It is not general error message. It is specific. And I have this error in an environment where the same record is changed from different points in the system. I have res_parent extended with two additional fields which values is changed from different points via REST API. This change is often in the same time. Often is different time but yet this error occurs because previous change is not committed yet. That's the issue, not offtopic. But Odoo code as whole is one big pile of crap and now it's clear to me why is that. It's because of Odoo S.A. extorting support fees, because of the Indian devs and the community which you proved not helpful. Thanks.

**Comment by odony** [comment]
Note: In general a concurrent update error is not a bug, it's a normal consequence of the optimistic locking mechanism, and we have mechanisms in place to auto-retry the failed transactions when it's possible and safe, e.g. in normal backend transactions. The alternative "pessimistic locking" is not an option because it completely kills performance.

In some cases however we may be able to entirely avoid the conflict, by making sure we don't write on "shared records". That makes sense when the conflict is *likely* and *avoidable*.
An example is when "ir.sequence" is used to provide sequence numbers. It is very conflict-prone because the same sequence is shared by all transactions on the same "journal". As a workaround we implement "standard" sequences as an alternative to "no-gap" sequences: when gaps in the numbers are acceptable, we avoid writing on the sequence record in order to entirely avoid the concurrent update. When gaps are not allowed (invoices) there is no workaround, the transactions must be processed sequentially in order to properly update the sequence journal (any transaction can fail until it is fully committed to database)

**Comment by pedrobaeza** [comment]
Thanks for the complementary information, @odony 

@bnisevic if you don't like Odoo by truly or misleading reasons, go to other software, but emitting that kind of offensive comments what you only get is to be banned and don't improve anything.

Look yourself if the problem is in Odoo or in your REST API, that don't retry the transaction when getting that message...

**Comment by odony** [comment]
@bnisevic you seem to be describing the expected behavior as well. Updating the same record from different transactions at the same time is unsafe and Odoo does not allow that. It's a critical feature to provide proper data integrity and avoid ["phantom reads" and "non-repeatable reads"](https://www.postgresql.org/docs/9.5/transaction-iso.html#MVCC-ISOLEVEL-TABLE), which are very real problems for an ERP. Hence the isolation level.
Without this it would be impossible to guarantee that stock levels are correct in a company with a high volume of logistics activity, for example.

You are always free to implement custom transactions with different levels of isolation in your own code, but you'd be in unsupported territory, and it's often not worth the trouble, because solutions exist with the default isolation.

If you can't simply work on different records, you could look into using a specific "NOWAIT" query to detect transaction conflicts early and process the transactions differently. That's what the `stock.quant` management system does to detect concurrent update conflicts ([here](https://github.com/odoo/odoo/blob/78004338969b996d6ca0b881691d60e74ff77a65/addons/stock/models/stock_quant.py#L214-L226)) and activate a "delayed quant merge" mode ([here](https://github.com/odoo/odoo/blob/78004338969b996d6ca0b881691d60e74ff77a65/addons/stock/models/stock_quant.py#L308-L337)), with almost the same speed and most importantly, the same integrity guarantees.

**Comment by bnisevic** [comment]
Thank you @odony! Your answer was very helpful. And it gives some clarification to the issue.

IMHO what is needed is some kind of mechanism within write() function where with a parameter one could set default behavior on the concurrent update error. Right now I can't rollback and repeat with write(), I can do that only with the cursor.

**Comment by odony** [comment]
@legalsylvain:

**tl;dr:** in order to fix your problem you must first identify the transaction that causes the conflict ;)

> For my side, it occurs when the create_from_ui (point_of_sale) function is called many times.

Who calls it, and for what session(s)? Normally there should be only one browser working one a given POS session at any time, so no chance of concurrent transactions.

> * Odoo 8.0 (OCB)

:see_no_evil: I guess you should be talking to the OCB maintainers for solving your issue, as 8.0 is EOL since 2017 :stuck_out_tongue: j/k


> TransactionRollbackError could not serialize access due to concurrent update
> bad query: UPDATE "account_bank_statement" SET "total_entry_encoding"='198.00' WHERE id = 54913

As we see the conflict occurs because 2 transactions are updating the same bank statement. For the POS, normally 1 browser = 1 POS config = 1 session = 1 bank statement. If multiple users/browsers use the same session, you know where the problem comes from ;-)

The auto-retry mechanism should normally take care of replaying a one-off concurrent update error, but if you're reporting it, I guess it's not sufficient. Perhaps more than 2 transactions are conflicting, and the auto-retry conflicts several times, eventually bailing out?

If you can confirm that only 1 browser is used for each POS session, then the conflict probably occurs with a different kind of transaction (not another `create_from_ui`). Could it be that someone is working on the POS session/orders from the backend while it's still open in the POS? Or working on the statement/statement lines, like [un]reconciling them?
You should be able to determine it by looking at logs around the time of the TransactionRollbackError, and seeing what other requests completed just before the error happens. This is likely the transaction that conflicted.

Typically a call to `create_from_ui()` can be rather long to process if there are many orders, and it must be committed all at once. Such a long transaction is more likely to be interrupted if a short transaction comes before, because the first commit wins.

**Comment by odony** [comment]
> IMHO what is needed is some kind of mechanism within write() function where with a parameter one could set default behavior on the concurrent update error. Right now I can't rollback and repeat with write(), I can do that only with the cursor.

The thing is that we can only replay a complete transaction, not a single `write()` (unless of the course the whole transaction was a RPC call to `some_model.write`), because we can only rollback an entire transaction by default. You can have a look at how the model dispatcher detects errors and automatically retries them in the [`@check` decorator](https://github.com/odoo/odoo/blob/d11956517be0a654b6772424420470b380bb81a1/odoo/service/model.py#L42-L143) that is used to protect all [RPC calls](https://github.com/odoo/odoo/blob/d11956517be0a654b6772424420470b380bb81a1/odoo/service/model.py#L155-L156).

In a custom REST controller your options could be:
- re-implement or call the generic retry mechanism of `@check` from `odoo.service.model.dispatch()`, which works transaction-wise.
- deal with it at the protocol level: do a FOR UPDATE NOWAIT query to detect concurrent updates before the `write()`, and return some well known value to the client to tell it to retry later
- (dirtier) make a simpler version of the retry thing, possibly breaking the atomicity of your transactions e.g. something like:
``` python
def _concurrent_write(self, vals):
    # flush any pending changes, in case of rollback
    # (breaks ACID!)
    self.env.cr.commit() 
    while True:
        try:
            # make the conflict-prone UPDATE and break if success
            self.write(vals)
            break
        except psycopg2.OperationalError as e:
            if e.code in odoo.service.model.PG_CONCURRENCY_ERRORS_TO_RETRY:
                # prepare for retry
                self.env.cr.rollback()
                time.sleep(.1)
            else:
                # don't hide non-concurrency errors
                raise
```
